ompi/communicator/comm_ft.c

```diff
@@ -0,0 +1,322 @@
+ * with this list via the MPI functions, and all failure notifications are reported
+ * in the context of a communicator. Thus, using a single index to know the user-level
+ * acknowledged failure is the simplest solution.
+ */
+int ompi_comm_failure_ack_internal(ompi_communicator_t* comm)
+{
+    /* Re-enable ANY_SOURCE */
+    comm->any_source_enabled = true;
+
+    /* Fix offset in the global failed list */
+    comm->any_source_offset = ompi_group_size(ompi_group_all_failed_procs);
+
+    return OMPI_SUCCESS;
+}
+
+int ompi_comm_failure_get_acked_internal(ompi_communicator_t* comm, ompi_group_t **group )
+{
+    int ret, exit_status = OMPI_SUCCESS;
+    int range[3];
+    ompi_group_t *tmp_sub_group = NULL;
+
+    /*
+     * If no failure present, then return the empty group
+     */
+    if( 0 == comm->any_source_offset ) {
+        *group = MPI_GROUP_EMPTY;
+        OBJ_RETAIN(MPI_GROUP_EMPTY);
+        exit_status = OMPI_SUCCESS;
+        goto cleanup;
+    }
+
+    tmp_sub_group = OBJ_NEW(ompi_group_t);
+
+    /*
+     * Access just the offset number of failures
+     */
+    range[0] = 0;
+    range[1] = comm->any_source_offset - 1;
+    range[2] = 1;
+
+    ret = ompi_group_range_incl(ompi_group_all_failed_procs, 1, &range, &tmp_sub_group);
+    if( OMPI_SUCCESS != ret ) {
+        exit_status = ret;
+        goto cleanup;
+    }
+
+    /*
+     * Access the intersection between the failed subgroup and our group
+     */
+    if( OMPI_COMM_IS_INTER(comm) ) {
+        ret = ompi_group_intersection(tmp_sub_group,
+                                      comm->c_local_group,
+                                      group);
+    } else {
+        ret = ompi_group_intersection(tmp_sub_group,
+                                      comm->c_remote_group,
+                                      group);
+    }
+
+    if( OMPI_SUCCESS != ret ) {
+        exit_status = ret;
+        goto cleanup;
+    }
+
+ cleanup:
+    if( NULL != tmp_sub_group ) {
+        OBJ_RELEASE(tmp_sub_group);
+        tmp_sub_group = NULL;
+    }
+
+    return exit_status;
+}
+
+int ompi_comm_shrink_internal(ompi_communicator_t* comm, ompi_communicator_t** newcomm)
+{
+    int ret, exit_status = OMPI_SUCCESS;
+    int flag = 1;
+    ompi_group_t *failed_group = NULL, *comm_group = NULL, *alive_group = NULL, *alive_rgroup = NULL;
+    ompi_communicator_t *newcomp = NULL;
+    int mode;
+    double start, stop;
+
+    *newcomm = MPI_COMM_NULL;
+
+    /*
+     * Step 1: Agreement on failed group in comm
+     */
+    /* --------------------------------------------------------- */
+    OPAL_OUTPUT_VERBOSE((5, ompi_ftmpi_output_handle,
+                         "%s ompi: comm_shrink: Agreement on failed processes",
+                         OMPI_NAME_PRINT(OMPI_PROC_MY_NAME) ));
+    start = MPI_Wtime();
+    ompi_group_intersection(comm->c_remote_group, ompi_group_all_failed_procs, &failed_group);
+    stop = MPI_Wtime();
```

Comment:
Timing is only used if OPAL_OUTPUT_VERBOSE with level 10 or above. 

Shouldn't the start/stop = MPI_Wtime() be protected with a macro?